src/main/frontend/fs/sync.cljs

```diff
@@ -0,0 +1,1223 @@
+                                     (if-let [v (poll! remote-changes-chan)]
+                                       (let [last-txid (:txid v)
+                                             current-txid (:txid data)]
+                                         (if (> last-txid current-txid)
+                                           (offer! remote-changes-chan v)
+                                           (offer! remote-changes-chan data)))
+                                       (offer! remote-changes-chan data))))))
+
+(defn ws-listen!
+  "return channal which output messages from server"
+  [graph-uuid *ws]
+  (let [remote-changes-chan (chan (async/sliding-buffer 1))]
+    (ws-listen!* graph-uuid *ws remote-changes-chan)
+    remote-changes-chan))
+
+(defn- get-json-body [body]
+  (or (and (map? body) body)
+      (or (string/blank? body) nil)
+      (js->clj (js/JSON.parse body) :keywordize-keys true)))
+
+(defn- get-resp-json-body [resp]
+  (-> resp (:body) (get-json-body)))
+
+(defn- request-once [api-name body token]
+  (go
+    (let [resp (http/post (str "https://api.logseq.com/file-sync/" api-name)
+                          {:oauth-token token
+                           :body (js/JSON.stringify (clj->js body))})]
+      {:resp (<! resp)
+       :api-name api-name
+       :body body})))
+
+(defn- request
+  ([api-name body token refresh-token-fn] (request api-name body token refresh-token-fn 0))
+  ([api-name body token refresh-token-fn retry-count]
+   (go
+     (let [resp (<! (request-once api-name body token))]
+       (if (and
+            (= 401 (get-in resp [:resp :status]))
+            (= "Unauthorized" (:message (get-json-body (get-in resp [:resp :body])))))
+         (do
+           (println "will retry after" (min 60000 (* 1000 retry-count)) "ms")
+           (<! (timeout (min 60000 (* 1000 retry-count))))
+           (let [token (<! (refresh-token-fn))]
+             (<! (request api-name body token refresh-token-fn (inc retry-count)))))
+         (:resp resp))))))
+
+(defn- remove-dir-prefix [dir path]
+  (let [r (string/replace path (js/RegExp. (str "^" dir)) "")]
+    (if (string/starts-with? r "/")
+      (string/replace-first r "/" "")
+      r)))
+
+(defn- remove-user-graph-uuid-prefix
+  "<user-uuid>/<graph-uuid>/path -> path"
+  [path]
+  (let [parts (string/split path "/")]
+    (if (and (< 2 (count parts))
+             (= 36 (count (parts 0)))
+             (= 36 (count (parts 1))))
+      (string/join "/" (drop 2 parts))
+      path)))
+
+(defn- encode-filepath
+  [filepath]
+  (->> (string/split filepath "/")
+       (remove empty?)
+       (map js/encodeURIComponent)
+       (string/join "/")))
+
+(defprotocol IRelativePath
+  (-relative-path [this]))
+
+(defprotocol IStoppable
+  (-stop! [this]))
+(defprotocol IStopped?
+  (-stopped? [this]))
+                                        ;from-path, to-path is relative path
+(deftype FileTxn [from-path to-path updated? deleted? txid]
```

Comment:
I'd normally recommend defrecord over deftype where possible given our average coworker's Clojure knowledge. Could you add a comment why you choose deftype over defrecord for these types?